### ISSUE TITLE
cannot enable Bandwidth Manager on oracle linux 9, 

### ISSUE THREAD

**Issue by heroes1412** [cilium/cilium#33455]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

new installed rke2 cluster without CNI
use helm to install cilium: helm upgrade --install cilium cilium/cilium --namespace kube-system --version 1.15.6  --set bandwidthManager.enabled=true

but when i check with:
kubectl -n kube-system exec ds/cilium -- cilium status | grep "BandwidthManager"

it shows "BandwidthManager:        Disabled"

### Cilium Version

1.15.6

### Kernel Version

Linux cp1 5.15.0-205.149.5.1.el9uek.x86_64 #2 SMP Fri Apr 5 11:29:36 PDT 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.28.10+rke2r1

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by joestringer** [comment]
Hi @heroes1412, the behaviour you report does sound surprising. To understand the issue better, we will need at least the cilium-agent logs from the environment when you experience this issue, or better a sysdump.

**Comment by heroes1412** [comment]
[cilium-sysdump-20240716-090436.zip](https://github.com/user-attachments/files/16242238/cilium-sysdump-20240716-090436.zip)
hi @joestringer , upload sysdump to you.

uname -a
Linux cp1 5.15.0-205.149.5.1.el9uek.x86_64 #2 SMP Fri Apr 5 11:29:36 PDT 2024 x86_64 x86_64 x86_64 GNU/Linux

i run helm upgrade cilium:
helm upgrade cilium cilium/cilium   --namespace kube-system   --reuse-values   --set bandwidthManager.enabled=true

[root@cp1 ~]# kubectl -n kube-system exec ds/cilium -- cilium status
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
KVStore:                 Ok   Disabled
Kubernetes:              Ok   1.28 (v1.28.10+rke2r1) [linux/amd64]
Kubernetes APIs:         ["EndpointSliceOrEndpoint", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "cilium/v2alpha1::CiliumCIDRGroup", "core/v1::Namespace", "core/v1::Pods", "core/v1::Service", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:    True   [ens160  192.168.2.33 (Direct Routing)]
Host firewall:           Disabled
SRv6:                    Disabled
CNI Chaining:            none
CNI Config file:         successfully wrote CNI configuration file to /host/etc/cni/net.d/05-cilium.conflist
Cilium:                  Ok   1.15.6 (v1.15.6-a09e05e6)
NodeMonitor:             Listening for events on 4 CPUs with 64x4096 of shared memory
Cilium health daemon:    Ok
IPAM:                    IPv4: 22/254 allocated from 172.16.0.0/24,
ClusterMesh:             0/0 clusters ready, 0 global-services
IPv4 BIG TCP:            Disabled
IPv6 BIG TCP:            Disabled
BandwidthManager:        Disabled
Host Routing:            BPF
Masquerading:            BPF   [ens160]   172.16.0.0/16 [IPv4: Enabled, IPv6: Disabled]
Controller Status:       120/120 healthy
Proxy Status:            OK, ip 172.16.0.39, 2 redirects active on ports 10000-20000, Envoy: embedded
Global Identity Range:   min 65536, max 131071
Hubble:                  Ok              Current/Max Flows: 4095/4095 (100.00%), Flows/s: 28.42   Metrics: Ok
Encryption:              Disabled

**Comment by heroes1412** [comment]
update, i already tried to install new rke2 cluster with cilium CNI on Ubuntu 24.04 LTS
root@rke2:~# uname -a
Linux rke2 6.8.0-35-generic #35-Ubuntu SMP PREEMPT_DYNAMIC Mon May 20 15:51:52 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

i use: "--set bandwidthManager.enabled=true --set bandwidthManager.bbr=true" but still cannot enable bandwidth manager, dont know why

[cilium-sysdump-20240716-102411-ubuntu24.04LTS.zip](https://github.com/user-attachments/files/16242791/cilium-sysdump-20240716-102411-ubuntu24.04LTS.zip)

**Comment by joestringer** [comment]
From the sysdump, when Cilium starts up it is not recognizing the ConfigMap configuration for this flag:

```
$ grep -R bandwidth-manager
logs-cilium-v7mh7-cilium-agent-20240716-090436-prev.log:2024-06-29T09:34:35.350327628+07:00 time="2024-06-29T02:34:35Z" level=info msg="  --enable-bandwidth-manager='false'" subsys=daemon
logs-cilium-v7mh7-cilium-agent-20240716-090436.log:2024-07-16T08:18:10.248583869+07:00 time="2024-07-16T01:18:10Z" level=info msg="  --enable-bandwidth-manager='false'" subsys=daemon
cilium-configmap-20240716-090436.yaml:  enable-bandwidth-manager: "true"
cilium-configmap-20240716-090436.yaml:        f:enable-bandwidth-manager: {}
cilium-bugtool-cilium-v7mh7-20240716-090436/cmd/cilium-debuginfo-20240716-020442.62+0000-UTC.md:enable-bandwidth-manager:false
cilium-bugtool-cilium-v7mh7-20240716-090436/cmd/cilium-debuginfo-20240716-020442.62+0000-UTC.json:    "bandwidth-manager": {
cilium-bugtool-cilium-v7mh7-20240716-090436/cmd/cilium-debuginfo-20240716-020442.62+0000-UTC.json:    "enable-bandwidth-manager:false",
```

**Comment by joestringer** [comment]
In my local kind environment, I can't quite reproduce the same behaviour with the latest v1.15.x release. The bandwidth manager is initially enabled, but then later disabled due to some issue detecting the default qdisc. This is unrelated to the issue you're observing.

```
$ helm install cilium cilium/cilium --version 1.15.7 \
  --namespace kube-system \
  --set bandwidthManager.enabled=true
NAME: cilium
LAST DEPLOYED: Tue Jul 16 10:44:22 2024
NAMESPACE: kube-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
You have successfully installed Cilium with Hubble.

Your release version is 1.15.7.

For any further help, visit https://docs.cilium.io/en/v1.15/gettinghelp
$ kubectl get ConfigMap cilium-config -n kube-system -o yaml | grep bandwidth
  enable-bandwidth-manager: "true"
$ kubectl -n kube-system logs pod/cilium-9h4kk | grep bandwidth
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
time="2024-07-16T17:44:56Z" level=info msg="  --enable-bandwidth-manager='true'" subsys=daemon
time="2024-07-16T17:44:57Z" level=warning msg="BPF bandwidth manager could not read procfs. Disabling the feature." error="Failed to read /host/proc/sys/net/core/default_qdisc: open /host/proc/sys/net/core/default_qdisc: no such file or directory" subsys=bandwidth-manager
```

**Comment by joestringer** [comment]
I don't understand why cilium-agent is not respecting the cilium-config ConfigMap in the initial environment from the sysdump. There seems to be a disconnect when the ConfigMap is synced to the local `/tmp` filesystem on the target node. You could try `grep -R bandwidth /tmp` to get some further information.The main difference I can see is that I am trying to reproduce using kind vs. the issue report is using rke2. Perhaps rke2 is somehow interfering with the configmap replication?

**Comment by joestringer** [comment]
In the new environment there is a log message describing why bandwidthManager is disabled:

```
BPF bandwidth manager could not detect host devices. Disabling the feature.
```

```
$ grep -R "bandwidth.*man" .
./logs-cilium-operator-98b4f6557-9fsk5-cilium-operator-20240716-102411.log:2024-07-16T10:18:07.562528606+07:00 time="2024-07-16T03:18:07Z" level=info msg="  --enable-bandwidth-manager='true'" subsys=cilium-operator-generic
./cilium-configmap-20240716-102411.yaml:  enable-bandwidth-manager: "true"
./cilium-configmap-20240716-102411.yaml:        f:enable-bandwidth-manager: {}
./logs-cilium-llfx2-cilium-agent-20240716-102411.log:2024-07-16T10:21:55.104193621+07:00 time="2024-07-16T03:21:55Z" level=info msg="  --enable-bandwidth-manager='true'" subsys=daemon
./logs-cilium-llfx2-cilium-agent-20240716-102411.log:2024-07-16T10:21:55.863661371+07:00 time="2024-07-16T03:21:55Z" level=warning msg="BPF bandwidth manager could not detect host devices. Disabling the feature." subsys=bandwidth-manager
./logs-cilium-llfx2-cilium-agent-20240716-102411.log:2024-07-16T10:21:55.863827365+07:00 time="2024-07-16T03:21:55Z" level=info msg="Start hook executed" duration="294.165µs" function="*bandwidth.manager.Start" subsys=hive
./cilium-bugtool-cilium-llfx2-20240716-102411/cmd/cilium-debuginfo-20240716-032413.703+0000-UTC.md:enable-bandwidth-manager:true
./cilium-bugtool-cilium-llfx2-20240716-102411/cmd/cilium-debuginfo-20240716-032413.703+0000-UTC.json:    "bandwidth-manager": {
./cilium-bugtool-cilium-llfx2-20240716-102411/cmd/cilium-debuginfo-20240716-032413.703+0000-UTC.json:    "enable-bandwidth-manager:true",
./cilium-bugtool-cilium-llfx2-20240716-102411/cmd/cilium-dbg-metrics-list.md:cilium_errors_warnings_total                             level="warning" subsystem="bandwidth-manager"                                                  
```

**Comment by heroes1412** [comment]
maybe "stable version" is not really stable :D

**Comment by joestringer** [comment]
I suspect what's going on here is that the device detection changed in Cilium 1.15. Hence the message about not being able to detect host devices. Maybe someone from @cilium/sig-datapath could help to provide steps about what a user could do if they encounter this error message.

**Comment by borkmann** [comment]
@hemanthmalla recently fixed a regression via https://github.com/cilium/cilium/pull/33161 . @heroes1412 Did you check with 1.16 as well?

**Comment by joamaki** [comment]
I think this is an issue with v1.15 still having the global `option.Config.devices` variable and that it's updated via `DeviceManager.Detect` which is called from `newDaemon`. From the logs it's clear that this happens after bandwidth-manager starts, hence it sees an empty list of devices.

In v1.16 this global variable is gone and these ordering issues are no longer there. Looks like for v1.15 we'll need a fix for this.
I'll take a look at reproducing this tomorrow and preparing a fix.

**Comment by julianwiedmann** [comment]
With https://github.com/cilium/cilium/pull/33910 merged, this should be fixed with the next patch release (v1.15.8).

**Comment by heroes1412** [comment]
fixed in v1.16.0 :D
